fdbserver/AuditUtils.actor.cpp

```diff
@@ -0,0 +1,187 @@
+/*
+ * AuditUtils.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbserver/AuditUtils.actor.h"
+
+#include "fdbclient/Audit.h"
+#include "fdbclient/FDBTypes.h"
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbserver/Knobs.h"
+
+#include "flow/actorcompiler.h" // has to be last include
+
+namespace {
+ACTOR Future<std::vector<AuditStorageState>> getLatestAuditStatesImpl(Transaction* tr, AuditType type, int num) {
+	state std::vector<AuditStorageState> auditStates;
+
+	loop {
+		auditStates.clear();
+		try {
+			RangeResult res = wait(tr->getRange(auditKeyRangeForType(type), num, Snapshot::False, Reverse::True));
+			for (int i = 0; i < res.size(); ++i) {
+				auditStates.push_back(decodeAuditStorageState(res[i].value));
+			}
+			break;
+		} catch (Error& e) {
+			wait(tr->onError(e));
+		}
+	}
+
+	return auditStates;
+}
+} // namespace
+
+ACTOR Future<UID> persistNewAuditState(Database cx, AuditStorageState auditState) {
+	state Transaction tr(cx);
+	state UID auditId;
+
+	loop {
+		tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+		try {
+			std::vector<AuditStorageState> auditStates = wait(getLatestAuditStatesImpl(&tr, auditState.getType(), 1));
+			uint64_t nextId = 1;
+			if (!auditStates.empty()) {
+				nextId = auditStates.front().id.first() + 1;
+			}
+			auditId = UID(nextId, deterministicRandom()->randomUInt64());
+			auditState.id = auditId;
+			tr.set(auditKey(auditState.getType(), auditId), auditStorageStateValue(auditState));
+			wait(tr.commit());
+			TraceEvent("PersistedNewAuditState", auditId).detail("AuditKey", auditKey(auditState.getType(), auditId));
+			break;
+		} catch (Error& e) {
+			wait(tr.onError(e));
+		}
+	}
+
+	return auditId;
+}
+
+ACTOR Future<Void> persistAuditState(Database cx, AuditStorageState auditState) {
+	state Transaction tr(cx);
+
+	loop {
+		tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+		try {
+			tr.set(auditKey(auditState.getType(), auditState.id), auditStorageStateValue(auditState));
```

Comment:
do we want to check the `auditKey` existed?

Reply:
Thanks for the tip, added some simple correctness check.